source/slang/slang-check-decl.cpp

```diff
@@ -2487,6 +2489,69 @@ namespace Slang
         return true;
     }
 
+    void SemanticsDeclHeaderVisitor::validateGenericConstraintSubType(GenericTypeConstraintDecl* decl, TypeExp type)
+    {
+        // Validate that the sub type of a constraint is in valid form.
+        //
+        if (auto subDeclRef = isDeclRefTypeOf<Decl>(type.type))
+        {
+            if (subDeclRef.getDecl()->parentDecl == decl->parentDecl)
+            {
+                // OK, sub type is one of the generic parameter type.
+                return;
+            }
+            if (as<GenericDecl>(decl->parentDecl))
+            {
+                // If the constraint is in a generic decl, then the sub type must be dependent on at least one
+                // of the generic type parameters defined in the same generic decl.
+                // For example, it is invalid to define a constraint like `void foo<T>() where int : float` since
+                // `int` isn't dependent on any generic type parameter.
+                auto dependentGeneric = getShared()->getDependentGenericParent(subDeclRef);
+                if (dependentGeneric.getDecl() != decl->parentDecl)
+                {
+                    getSink()->diagnose(type.exp, Diagnostics::invalidConstraintSubType, type);
+                    return;
+                }
+            }
+            else if (as<AssocTypeDecl>(decl->parentDecl))
+            {
+                // If the constraint is on an associated type, then it should either be the associated type itself,
+                // or a associated type of the associated type.
+                // For example,
+                // ```
+                // interface IFoo {
+                //    associatedtype T
+                //        where T : IFoo  // OK, constraint is on the associatedtype T itself.
+                //        where T.T == X; // OK, constraint is on the associated type of T.
+                //        where int == X; // Error, int is not a valid left hand side of a constraint.
```

Comment:
I think you meant to say 
```
where int : X // Error ...
```
I mean that `==` should be replaced with `:` and `;` should be dropped.

Reply:
`==` is also a valid constraint operator, for type equality constraints.